experimental/ast/token.go

```diff
@@ -0,0 +1,641 @@
+	_TokenUnused // Reserved for future use.
+
+	// DO NOT ADD MORE TOKEN KINDS: ONLY THREE BITS ARE AVAILABLE
+	// TO STORE THEM.
+)
+
+// TokenKind identifies what kind of token a particular [Token] is.
+type TokenKind byte
+
+// IsSkippable returns whether this is a token that should be examined during
+// syntactic analysis.
+func (t TokenKind) IsSkippable() bool {
+	return t == TokenSpace || t == TokenComment || t == TokenUnrecognized
+}
+
+// String implements [strings.Stringer] for TokenKind.
+func (t TokenKind) String() string {
+	switch t {
+	case TokenUnrecognized:
+		return "TokenUnrecognized"
+	case TokenSpace:
+		return "TokenSpace"
+	case TokenComment:
+		return "TokenComment"
+	case TokenIdent:
+		return "TokenIdent"
+	case TokenString:
+		return "TokenString"
+	case TokenNumber:
+		return "TokenNumber"
+	case TokenPunct:
+		return "TokenPunct"
+	default:
+		return fmt.Sprintf("TokenKind(%d)", int(t))
+	}
+}
+
+// Token is a lexical element of a Protobuf file.
+//
+// Protocompile's token stream is actually a tree of tokens. Some tokens, called
+// non-leaf tokens, contain a selection of tokens "within" them. For example, the
+// two matched braces of a message body are a single token, and all of the tokens
+// between the braces are contained inside it. This moves certain complexity into
+// the lexer in a way that allows us to handle matching delimiters generically.
+//
+// The zero value of Token is the so-called "nil token", which is used to denote the
+// absence of a token.
+type Token struct {
+	withContext
+
+	raw rawToken
+}
+
+// IsPaired returns whether this is a non-nil leaf token.
+func (t Token) IsLeaf() bool {
+	if t.Nil() {
+		return false
+	}
+
+	if impl := t.impl(); impl != nil {
+		return impl.IsLeaf()
+	}
+	return t.synthetic().IsLeaf()
+}
+
+// IsSynthetic returns whether this is a non-nil synthetic token (i.e., a token that didn't
+// come from a parsing operation.)
+func (t Token) IsSynthetic() bool {
+	return t.raw < 0
+}
+
+// Kind returns what kind of token this is.
+//
+// Returns [TokenUnrecognized] if this token is nil.
+func (t Token) Kind() TokenKind {
+	if t.Nil() {
+		return TokenUnrecognized
+	}
+
+	if impl := t.impl(); impl != nil {
+		return impl.Kind()
+	}
+	return t.synthetic().kind
+}
+
+// Text returns the text fragment referred to by this token.
+// Note that this DOES NOT include any child tokens!
+//
+// Returns empty string fot the nil token.
+func (t Token) Text() string {
+	if t.Nil() {
+		return ""
+	}
+
+	if synth := t.synthetic(); synth != nil {
+		if synth.kind == TokenString {
+			// If this is a string, we need to add quotes and escape it.
+			// This can be done on-demand.
+
+			var escaped strings.Builder
+			escaped.WriteRune('"')
+			for _, r := range synth.text {
+				switch {
+				case r == '\n':
+					escaped.WriteString("\\n")
+				case r == '\r':
+					escaped.WriteString("\\r")
+				case r == '\t':
+					escaped.WriteString("\\t")
```

Comment:
Let's also use short single-letter escapes for the other runes that Protobuf string literals support: `\a`, `\b`, `\f`, and `\v`.